src/main/java/io/jenkins/plugins/pipeline_elasticsearch_logs/write/fluentd/FluentdWriter.java

```diff
@@ -0,0 +1,268 @@
+
+    public String getTag() {
+        return tag;
+    }
+
+    @DataBoundSetter
+    public void setTag(String tag) {
+        this.tag = tag;
+        this.configChanged = true;
+    }
+
+    public String getHost() {
+        return host;
+    }
+
+    @DataBoundSetter
+    public void setHost(String host) {
+        this.host = host;
+        this.configChanged = true;
+    }
+
+    public int getPort() {
+        return port;
+    }
+
+    @DataBoundSetter
+    public void setPort(int port) {
+        if(port < 0 || port > 65535) throw new IllegalArgumentException("Port out of range: " + port);
+        this.port = port;
+        this.configChanged = true;
+    }
+
+    public int getRetryMillis() {
+        return retryMillis;
+    }
+
+    @DataBoundSetter
+    public void setRetryMillis(int retryMillis) {
+        if (retryMillis < 0)
+            retryMillis = 0;
+        this.retryMillis = retryMillis;
+        this.configChanged = true;
+    }
+
+    public int getTimeoutMillis() {
+        return timeoutMillis;
+    }
+
+    @DataBoundSetter
+    public void setTimeoutMillis(int timeoutMillis) {
+        if(timeoutMillis < 0) throw new IllegalArgumentException("timeoutMillis less than 0");
+        this.timeoutMillis = timeoutMillis;
+        this.configChanged = true;
+    }
+
+    public int getBufferCapacity() {
+        return bufferCapacity;
+    }
+
+    @DataBoundSetter
+    public void setBufferCapacity(int bufferCapacity) {
+        if (bufferCapacity < 0)
+            bufferCapacity = DEFAULT_BUFFER_CAPACITY;
+        this.bufferCapacity = bufferCapacity;
+        this.configChanged = true;
+    }
+
+
+    @Extension
+    @Symbol("fluentd")
+    public static class DescriptorImpl extends ElasticSearchWriteAccessDescriptor {
+        @Override
+        public String getDisplayName() {
+            return "Fluentd Elasticsearch Writer";
+        }
+    }
+
+    /**
+     * Logs the given object to Fluentd. This method blocks if Fluentd is not reachable and
+     * throws an IOException after TIMEOUT_MILLIS.
+     *
+     * @param data The data to post
+     * @throws IOException if something failed permanently
+     */
+    @Override
+    public void push(Map<String, Object> data) throws IOException {
+        if(fluentd == null || configChanged) initFluentdLogger();
+        try {
+
+            long timeMarker = System.currentTimeMillis();
+            boolean sent = fluentd.log(tag, data);
+            if(!sent) {
+                //Errors are not reported explicitly. But the latest error should be the reason for (sent == false)
+                Exception lastError = errorHandler.getLastError(timeMarker);
+                throw new IOException("Could not send logs to fluentd", lastError);
+            }
+
+            // (sent == true) does not necessarily mean the data has already been sent to fluentd.
+            // To ensure data can be sent we flush the library buffer every time.
+            flushWithRetry();
```

Comment:
Might slow down throughput unnecessarily (e.g. temporary unavailable fluentd), but I'm also of the opinion that ensuring the logs are sent is more important than throughput.

Reply:
Normally the flushWithRetry will do nothing as the data was already sent with the call of the log method.
I'm in favor of avoiding this method as a short unavailability of fluentd longer than the timeout (default is 3s) would lead to a build abort. As soon as fluentd becomes available again, all pending data is sent.
When we keep this we should at least set a rather long timeout.

Reply:
@anbrsap What do you think?

Reply:
Didn't look into the code. In general I think there should be a relatively small send buffer per build. A log call appends to the buffer and blocks as long as the buffer is full. The buffer gets flushed asynchronously. Having only a _small_ buffer prevents that a build progresses too much without having the respective logs in the log store or forwarder. In addition, it must be ensured that at the very end of a build the log buffer gets flushed (with retry and configurable timeout).

Regarding oversized log entries, I would recommend that the logger breaks them into multiple log events. This must happen in the method that writes into the log buffer.

Reply:
> In addition, it must be ensured that at the very end of a build the log buffer gets flushed (with retry and configurable timeout).

Important point. That has been one reason in the past why we/I flushed more often than required. Can we react on "end of pipeline" properly?

Reply:
The sender has an explicit close method we can use to flush the buffers. Note that almost every step creates its own sender instance that is closed when the step is finished.

Reply:
> 
> Regarding oversized log entries, I would recommend that the logger breaks them into multiple log events. This must happen in the method that writes into the log buffer.

This might be complicated if we have annotations in the logs and want to keep them in the events . Also we would need something on the UI side reading the logs that will merge back the events that were too big.

Reply:
> > Regarding oversized log entries, I would recommend that the logger breaks them into multiple log events. This must happen in the method that writes into the log buffer.
> 
> This might be complicated if we have annotations in the logs and want to keep them in the events . Also we would need something on the UI side reading the logs that will merge back the events that were too big.

If a single annotation is larger than the maximum log entry size, we must remove it even if annotations are to be kept.

Reassembly of fragmented log entries I regard as optional. We may add fragmentation info as additional fields of log entries. If a client doesn't do reassembly, the fragments events will appear as separate lines/entries.

Reply:
A more advanced option would be to evaluate the annotations, close them, open them again in the next fragment (where possible).
But I would also vote for simply removing them for now.